### PR TITLE
feat: fetch and expose manifest

### PR DESCRIPTION
**Pull request by khairul-alam-licon**
# Description
- Update the SDK API request to match the new API spec for retrieving the mini app manifest
- In the Demo App: Update the "first-time launch" screen to display the key/value data from the manifest, there are some UI tasks related to required and optional permissions left, I'll address those in a separate PR using MINI-2503

## Links
- MINI-2500

# Checklist
- [x] I have read the [contributing guidelines](../.github/CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](../.github/CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt [hunk 1]
```diff
@@ -123,6 +125,29 @@ internal class MiniAppDownloader(
         versionId: String
     ) = apiClient.fetchFileList(appId, versionId)
 
+    @Throws(MiniAppSdkException::class)
+    suspend fun fetchMiniAppManifest(appId: String, versionId: String): MiniAppManifest {
+        // TODO: fetchMiniAppManifest(appId, versionId)
+        if (versionId.isEmpty()) throw MiniAppSdkException("Provided Mini App Version ID is invalid.")
```
Comment (khairul-alam-licon) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt:131]: Is there any other criteria to define an invalid version id? 🤔

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt [hunk 2]
```diff
@@ -79,6 +80,32 @@ internal class ApiClient @VisibleForTesting constructor(
...
+            versionId = versionId,
+            testPath = testPath
+        )
+        return requestExecutor.executeRequest(request)
+    }
+
+    // TODO: Use fetchMiniAppManifest, and remove this function
+    @Suppress("ForbiddenComment")
+    @Throws(MiniAppSdkException::class)
+    fun fetchMockManifest(miniAppId: String, versionId: String): MetadataEntity {
```
Comment (khairul-alam-licon) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt:97]: TODO: will be removed later after api endpoint available.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt [hunk 3]
```diff
@@ -1,25 +1,40 @@
...
 
 private const val DEFAULT_ACCEPTANCE = false
 
 class PreloadMiniAppWindow(private val context: Context, private val preloadMiniAppLaunchListener: PreloadMiniAppLaunchListener) {
+    private val miniApp = MiniApp.instance(AppSettings.instance.miniAppSettings)
     private lateinit var preloadMiniAppAlertDialog: AlertDialog
     private lateinit var preloadMiniAppLayout: View
     private var miniAppInfo: MiniAppInfo? = null
     private var miniAppId: String = ""
+    private var miniAppVersionId: String = ""
```
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt:37]: version id is empty when trying to download a miniapp using only app-id from this page. 

Screen Shot 2021-02-15 at 14 21 05

Then metadata's permission value will not be found in first time UI. 
@corycaywood Do you think, this will be expected?

**Review by khairul-alam-licon — COMMENTED** [review]
Self review.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        @Path("hostappId") hostAppId: String,
+        @Path("miniappId") miniAppId: String,
+        @Path("versionId") versionId: String,
+        @Path("testPath") testPath: String = ""
+    ): Call<MetadataEntity>
+}
+
+@Keep
+internal data class MetadataEntity(
+    @SerializedName("metadata") val metadata: MetadataResponse
```
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:21]: I'm not sure if we can `MetadataResponse` directly. The Host App wants to have access to all key/values of the manifest through `MiniAppManifest.manifest`. But that does make things more complicated because we would need to parse this as `Map<String, String>`...

Maybe we should re-think our approach to custom keys for the Host App. Instead of giving access to the whole manifest through `MiniAppManifest.manifest`, could we add a key like `metadata` where the Host App can use any custom keys?

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt (hunk 3 above)
Comment (corycaywood) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt:37]: This one is supposed to launch the latest version ID for the specified mini app. So I think it needs to first call `MiniApp.fetchInfo` to get the latest version ID and then fetch the meta data for that version.

**Review by corycaywood — COMMENTED** [review]
Looking good 👍  See my comments below.

Just a note about the demo app UI (I'm not sure if you're planning to address that here or another PR):
See the iOS Demo App UI for the required/optional permissions. The user shouldn't be able to disable the required permissions - they just have the text `Required` in red without a toggle switch.

**Comment by khairul-alam-licon** [comment]
> Just a note about the demo app UI (I'm not sure if you're planning to address that here or another PR):
See the iOS Demo App UI for the required/optional permissions. The user shouldn't be able to disable the required permissions - they just have the text Required in red without a toggle switch.

Yeah, I'll address this in a separate PR using MINI-2503 ticket.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ */
+@Suppress("DataClassContainsFunctions")
+@Keep
+data class MiniAppManifest(
+    // List of permissions requested by Mini App in their manifest
+    val requiredPermissions: List<MiniAppCustomPermissionType>?,
+    val optionalPermissions: List<MiniAppCustomPermissionType>?,
+
+    // Map containing all key/values from the Mini App manifest
+    val manifest: Map<String, String>
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt:17]: Yes, I was about to ask from where to set this value 🤔

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import androidx.annotation.Keep
+import com.google.gson.annotations.SerializedName
+import retrofit2.Call
+import retrofit2.http.GET
+import retrofit2.http.Path
+
+internal interface MetadataApi {
+    @GET("host/{hostappId}/miniapp/{miniappId}/version/{versionId}/{testPath}/metadata")
+    fun fetchMetadata(
+        @Path("hostappId") hostAppId: String,
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:12]: Yeah, but I was checking in other places e.g. `AppInfoApi` / `ManifestApi`, we are still using `hostappId`, so I thought that is the consistent parameter. Although, we can wait for the metadata api endpoint release. I'll update accordingly.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import com.rakuten.tech.mobile.miniapp.permission.MiniAppCustomPermissionType
+
+/**
+ * A data class to represent data in the mini app's manifest.
+ */
+@Suppress("DataClassContainsFunctions")
+@Keep
+data class MiniAppManifest(
+    // List of permissions requested by Mini App in their manifest
+    val requiredPermissions: List<MiniAppCustomPermissionType>?,
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt:13]: Yeah, I'll update this part to support `reason` string.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,16 +1,23 @@
...
+- **Feature:** Added first-time launching screen before downloading any miniapp.
 
-`rasAppId` -> `rasProjectId`.
+### 3.X.X (In progress)
 
-`MiniAppMessageBridge.requestPermission` -> `MiniAppMessageBridge.requestDevicePermission`.
+- **Removed:** Cleanup deprecated components before v3.0.0. Please replace usages in your code as follows:
 
-Replace `getUserName`, `getProfilePhoto` with new prototypes.
+Before v3.0.0  |  Replace with v3.0.0
```
Comment (minh-rakuten) [inline CHANGELOG.md:14]: Nice table 👍 
`Replace with v3.0.0` can be `v3.0.0`.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapplist/MiniAppListFragment.kt [hunk 9]
```diff
@@ -43,7 +43,7 @@ class MiniAppListFragment : BaseFragment(), MiniAppListener, OnSearchListener,
     private lateinit var binding: MiniAppListFragmentBinding
     private lateinit var miniAppListAdapter: MiniAppListAdapter
     private lateinit var searchView: SearchView
-    private val preloadMiniAppWindow by lazy { PreloadMiniAppWindow(context!!, this) }
+    private val preloadMiniAppWindow by lazy { PreloadMiniAppWindow(requireContext(), this) }
```
Comment (minh-rakuten) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapplist/MiniAppListFragment.kt:46]: It's the same but `context!!` gives better minor performance :)

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt [hunk 10]
```diff
@@ -70,13 +86,48 @@ class PreloadMiniAppWindow(private val context: Context, private val preloadMini
...
+            .addItemDecoration(
+                DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
+            )
+
+        // TODO: inflate UI from MiniApp.getRequiredCustomPermissions
+        // TODO: inflate UI from MiniApp.getOptionalCustomPermissions
+        val namesForAdapter: ArrayList<MiniAppCustomPermissionType> = arrayListOf()
+        val resultsForAdapter: ArrayList<MiniAppCustomPermissionResult> = arrayListOf()
+
+        GlobalScope.launch(Dispatchers.IO) {
```
Comment (minh-rakuten) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt:105]: Can we do `getMiniAppManifest` in `ViewModel` where the container call, i.e `MiniAppListViewModel` and then pass the required parameters to `PreloadMiniAppWindow`?

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt [hunk 11]
```diff
@@ -1,25 +1,40 @@
...
+import com.rakuten.tech.mobile.testapp.ui.settings.AppSettings
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
 import java.lang.Exception
 
 private const val DEFAULT_ACCEPTANCE = false
 
 class PreloadMiniAppWindow(private val context: Context, private val preloadMiniAppLaunchListener: PreloadMiniAppLaunchListener) {
+    private val miniApp = MiniApp.instance(AppSettings.instance.miniAppSettings)
```
Comment (minh-rakuten) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt:32]: It's better not to call it here. It should be in the `ViewModel` of the view containing `PreloadMiniAppWindow`.

**Review by minh-rakuten — COMMENTED** [review]
Looks good.

I have comments on the api call that should not be in view component.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (corycaywood) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt:17]: See my comment below about MetadataResponse - I think we can replace this with something like `extras` or `customMetaData`. Basically a key where the Mini App can set any extra values required by the Host App:
```
// manifest.json
{
  reqPermissions: [],
  optPermissions: [],
  customMetaData: {
    randomTestKey: "value"
  }
}
```

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (corycaywood) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:12]: Hmm, I think we could change `hostappId` in the other classes too. Maybe it doesn't have to be in this PR - but it's up to you.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapplist/MiniAppListFragment.kt (hunk 9 above)
Comment (khairul-alam-licon) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapplist/MiniAppListFragment.kt:46]: Yeah still the same context, just wanted to avoid any possibility of NPE.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt]: Yeah, `hostappId` is deprecated-like. I'll use `hostId` in this line, and take a look in other places in a separate PR.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapplist/MiniAppListFragment.kt (hunk 9 above)
Comment (minh-rakuten) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapplist/MiniAppListFragment.kt:46]: Why do you want to avoid NPE? `requireContext()` is throwing `IllegalStateException` so I think they are same in exception throwing.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt (hunk 10 above)
Comment (khairul-alam-licon) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt]: Yeah, I have updated with `ViewModel`.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 4 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:22]: Yeah, currently I have something in this part with `customMetaData`, but I doubt that is not enough useful for hostapp.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    val optionalPermissions: List<MiniAppCustomPermissionType>?,
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt]: Agreed, I have updated this part.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 4 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:21]: I have updated this part.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt]: I have updated this part. Thanks for the idea.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+@Keep
+data class MiniAppManifest(
+    // List of permissions requested by Mini App in their manifest
+    val requiredPermissions: List<Pair<MiniAppCustomPermissionType, String>>,
+    val optionalPermissions: List<Pair<MiniAppCustomPermissionType, String>>,
+    // custom metadata set by Mini App in their manifest
+    private val customMetaData: Map<String, String>
+) {
+    // get metadata value of "customMetaData" per key set by Host App
+    fun getMetadataValue(key: String) = customMetaData[key]
```
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppManifest.kt:18]: Sorry, I think there was something like this in the draft specs, but it seems like we don't really need it now since we changed to using a specific key for custom meta data. 

Probably you can remove this and change `customMetaData` to public. Since you're using `Map` as the type it will already be immutable.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        @Path("hostId") hostId: String,
+        @Path("miniappId") miniAppId: String,
+        @Path("versionId") versionId: String,
+        @Path("testPath") testPath: String = ""
+    ): Call<MetadataEntity>
+}
+
+@Keep
+internal data class MetadataEntity(
+    @SerializedName("metadata") val metadata: MetadataResponse
```
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:21]: Seems that the Demo App is crashing when the `metadata` field isn't present. It's strange because it should instead throw an error which is caught by the Host App (a `MiniAppSdkException` type error)

**Review by corycaywood — COMMENTED** [review]
Looking really close 👍  See my comments below.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/permission/MiniAppPermission.kt [hunk 15]
```diff
@@ -22,9 +22,11 @@ enum class MiniAppCustomPermissionType(val type: String) {
     LOCATION("rakuten.miniapp.device.LOCATION"),
     UNKNOWN("UNKNOWN");
 
-    internal companion object {
+    companion object {
 
-        internal fun getValue(type: String) = values().find { it.type == type } ?: UNKNOWN
+        /** get the value of custom permission per type. **/
+        fun getValue(type: String): MiniAppCustomPermissionType =
```
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/permission/MiniAppPermission.kt:28]: Why do we need to make it public here? I think only miniapp sdk use it so that's why we made it internal.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 14 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt]: Yeah, looks like its a NPE has been caught because `MetadataResponse` is non-null here. I have changed it to nullable and handle the exception in `MiniAppDownloader.prepareMiniAppManifest`. Sorry, I didn't think about it before 😢

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        @Path("hostId") hostId: String,
+        @Path("miniappId") miniAppId: String,
+        @Path("versionId") versionId: String,
+        @Path("testPath") testPath: String = ""
+    ): Call<MetadataEntity>
+}
+
+@Keep
+internal data class MetadataEntity(
+    @SerializedName("metadata") val metadata: MetadataResponse?
```
Comment (khairul-alam-licon) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:21]: TODO: rename to `bundleManifest`.

File: miniapp/USERGUIDE.md [hunk 17]
```diff
@@ -445,7 +445,58 @@ CoroutineScope(Dispatchers.IO).launch {
...
+         {
+            "name":"rakuten.miniapp.user.CONTACT_LIST",
+            "reason":"Describe your reason here."
+         },
+         {
+            "name":"rakuten.miniapp.device.LOCATION",
+            "reason":"Describe your reason here."
+         }
+      ],
+      "customMetaData":{
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply miniapp/USERGUIDE.md]: Yeah, I have added an explanation on top of the json.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt [hunk 18]
```diff
@@ -123,6 +125,33 @@ internal class MiniAppDownloader(
...
+        try {
+            val requiredPermissions = metadataEntity.metadata?.requiredPermissions?.map {
+                Pair(MiniAppCustomPermissionType.getValue(it.name), it.reason)
+            } ?: emptyList()
+
+            val optionalPermissions = metadataEntity.metadata?.optionalPermissions?.map {
+                Pair(MiniAppCustomPermissionType.getValue(it.name), it.reason)
+            } ?: emptyList()
+
+            val customMetadata = metadataEntity.metadata?.customMetaData!!
```
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt:148]: I don't think we should use the not-null assertion here. The mini app might not have this key, so we shouldn't throw an error in that case. Also, `manifest.json` is technically optional, so it's possible for the mini app to have an empty manifest.

Maybe this instead:
```suggestion
            val customMetadata = metadataEntity.metadata?.customMetaData ?: emptyMap()
```

**Review by corycaywood — COMMENTED** [review]
Looks good 👍  Just one issue below, and also need to update to use the `bundleManifest` key.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    val manifestErrorData: LiveData<String>
+        get() = _manifestErrorData
+    val versionIdErrorData: LiveData<String>
+        get() = _versionIdErrorData
+
+    fun getMiniAppManifest(miniAppId: String, versionId: String, metadataKey: String) =
+        viewModelScope.launch(Dispatchers.IO) {
+            try {
+                val miniAppManifest = miniApp.getMiniAppManifest(miniAppId, versionId)
+                _miniAppManifest.postValue(miniApp.getMiniAppManifest(miniAppId, versionId))
```
Comment (corycaywood) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt:38]: Why do you need to call `miniApp.getMiniAppManifest` here again? Can't you just pass `miniAppManifest` from abov?

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+                _miniAppManifestMetadata.postValue(miniAppManifest.customMetaData[metadataKey])
```
Comment (corycaywood) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt:39]: I think we should display the whole `customMetaData` JSON here, not just a specific string. So maybe an easy solution is to just use `toString` or `Gson`?
```suggestion
                val metadata = GsonBuilder().setPrettyPrinting().create().toJson(miniAppManifest.customMetaData)
                _miniAppManifestMetadata.postValue(metadata)
```

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+                    if (versionId.isEmpty()) {
+                        miniAppVersionId.observe(lifecycleOwner, Observer { versionId = it })
+                        versionIdErrorData.observe(lifecycleOwner, Observer {
+                            Toast.makeText(context, it, Toast.LENGTH_LONG).show()
+                        })
+                    }
+
+                    miniAppManifestMetadata.observe(lifecycleOwner, Observer {
+                        preloadMiniAppLayout.findViewById<TextView>(R.id.preloadMiniAppMetaData).text =
+                            "Metadata value: " + it
```
Comment (corycaywood) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppWindow.kt:113]: Should this be a string resource instead? Also, I think text like `Custom MetaData` might be better to match the naming in `manifest.json`.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 16 above)
Comment (corycaywood) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:21]: I think you can go ahead and change this to `bundleManifest` so we can get the PR merged. We can make any minor changes later if needed.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt [hunk 22]
```diff
@@ -123,6 +125,33 @@ internal class MiniAppDownloader(
...
+            } ?: emptyList()
+
+            val optionalPermissions = metadataEntity.metadata?.optionalPermissions?.map {
+                Pair(MiniAppCustomPermissionType.getValue(it.name), it.reason)
+            } ?: emptyList()
+
+            val customMetadata = metadataEntity.metadata?.customMetaData ?: emptyMap()
+            return MiniAppManifest(requiredPermissions, optionalPermissions, customMetadata)
+        } catch (exception: NullPointerException) {
+            throw MiniAppSdkException("Metadata isn't found in manifest.json for this Mini App.")
```
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt:151]: I think you don't need this try/catch anymore, do you? Is there anything that could throw `NullPointerException`?

**Review by corycaywood — APPROVED** [review]
Just a few more comments and I think it should be good to merge.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt (hunk 19 above)
Comment (khairul-alam-licon) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt:38]: Ah, I don't know why I wrote like this! Sorry, I am updating.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 16 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:21]: Yeah, I can address it in a separate PR.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt (hunk 22 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt]: Yeah, now we can remove try/catch. I have updated this part.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt (hunk 20 above)
Comment (khairul-alam-licon) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/preload/PreloadMiniAppViewModel.kt]: This is a nice idea 👍

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt (hunk 16 above)
Comment (corycaywood) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/MetadataApi.kt:21]: I think you should change it in this PR to `bundleManifest` because that is supposed to be the correct format. It will potentially avoid the need for another PR.

Sorry, I meant that if there are any _unexpected_ changes/issues later, then we can make a separate PR.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt [hunk 23]
```diff
@@ -79,6 +80,17 @@ internal class ApiClient @VisibleForTesting constructor(
         return requestExecutor.executeRequest(request)
     }
 
+    @Throws(MiniAppSdkException::class)
+    suspend fun fetchMiniAppManifest(miniAppId: String, versionId: String): MetadataEntity {
```
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt:84]: We can have unit test here.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt [hunk 24]
```diff
@@ -124,6 +126,29 @@ internal class MiniAppDownloader(
...
+    suspend fun fetchMiniAppManifest(appId: String, versionId: String): MiniAppManifest {
+        if (versionId.isEmpty()) throw MiniAppSdkException("Provided Mini App Version ID is invalid.")
+        else {
+            val manifestResponse = apiClient.fetchMiniAppManifest(appId, versionId)
+            return prepareMiniAppManifest(manifestResponse)
+        }
+    }
+
+    @VisibleForTesting
+    fun prepareMiniAppManifest(metadataEntity: MetadataEntity): MiniAppManifest {
```
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt:139]: We need to have test cases for `metadata`, `requiredPermissions`, `optionalPermissions`, `customMetaData` when they are null/empty list

**Review by minh-rakuten — APPROVED** [review]
There are still remain works for unit tests. It's better to handle them in this PR but it's okay to do later, please take a note below.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt (hunk 23 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt:84]: Thanks, I think I have added testcase for this. I'll recheck in next PR.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt (hunk 24 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt:139]: Okay, I'll address this comment in next PR.

**Comment by khairul-alam-licon** [comment]
Thanks for the reviews guys.
